### DISCUSSION TITLE
[TPA] el validador de egresos no me trae los nuevos egresos

### DISCUSSION THREAD

**Discussion by foviedo** [dds-jv/foro#158] [Consultas de herramientas] [answered]
Buenas, qué tal? 
Con mi grupo tenemos un problema en el TP:
Usando el Scheluder at fixed rate, corremos una función que valida los egresos SIN_VALIDAR cada X tiempo, esto lo hace haciendo una query a la base de datos cada vez
Resulta que si yo ingreso un nuevo egreso desde al UI, estos nuevos que ingreso no se estarían trayendo desde la base de datos
Me fijé por el MySQL Workbench si efectivamente se estaba persistiendo todo bien, y sip, lo estaba haciendo, pero en el job no me los estaba trayendo, sólo lo hacía recién cuando paraba la ejecución y volvía a darle play. Lo que hago cada ese X tiempo es.
>		RepositorioEgresos.getInstance().egresosSinValidar().forEach(Egreso::validarme);

Y lo que hago ahí dentro de egresosSinValidar es:

> 		List<Egreso> losEgresos = entityManager().createQuery("from Egreso where estadoValidacion = 'SIN_VALIDAR'",Egreso.class).getResultList();
>		losEgresos.forEach(unEgreso -> unEgreso.instanciaTuValidador());
>		return losEgresos;

Este entityManager() es el de WithGlobalEntityManager()

Pensamos en que tal vez tenía que ser un problema similar al de #73, así que cambiamos al global por un PerThreadEntityManager y lo cerrábamos luego. Esto nos solucionó parcialmente el problema, pues surgió otro que esta vez sí no entendemos por qué ocurre
Básicamente, se valida bien, pero lo que ocurre, es que cuando ponemos la página de "Ver Presupuestos" el último presupuesto que insertamos, a veces aparece y a veces no (si recargo la página aparece, la recargo y vuelve a no aparecer y así)

Es muy raro realmente, alguien tiene idea de qué puede llegar a ser? En caso de ser necesario, puedo detallar mejor el código
Saludos,
Facundo

EDIT: Dicho sea de paso, si no hago esto de correr cada X tiempo las validaciones, anda todo bien

**Comment by gastonprieto** [comment]
Definitivamente tiene que ver con https://github.com/dds-jv/foro/issues/73.

En el job tienen que borrar/limpiar las cosas que buscaron porque sino quedan cachedas.
Y en la UI también tienen que hacerlo en los after, sino lo hacen va a pasar ese comportamiento erratico

También es importante manejar las transacciones, porque si no se confirman no se va a ver, te dejo mas info en este post: https://groups.google.com/g/dds-jv-cursada/c/MR4cvLUY1mw/m/cP4MHXZGDAAJ

cualquier cosa avisanos

**Comment by foviedo** [comment, marked as answer]
Gastón no voy a mentirte, me puse a buscar en el código qué cosas me faltaron hacerlas con transaction/abriendo y cerrando el entityManager y el tp ahora anda bien sin que haya hecho ningún cambio. Así que digamos que un hechicero lo hizo
Gracias!
